### PR TITLE
Callback PR Rev 3

### PR DESCRIPTION
**Pull request by blisc**
Rebased on top of Master. Replaces #597 

Changelog:
Major:

- Reworked the callbacks systems to be more user-friendly
- Added new callbacks: 
    - `SimpleLogger` and `TensorboardLogger` which replace `SimpleLossLoggerCallback`
    - `WandBLogger` replaces `WandbCallback`
    - `CheckpointCallback` has been updated to the new callback system but usage and functionality remains the same as before.
- Remove the logic of `__get_top_sorted_modules_and_dataloader` and split it off into a new function `topological_sort_from_leaves` which now lives in core/neural_factory.py
    - It could be moved back into backends/pytorch/actions.py, might be a better idea
- Removed PtActions.modules in favour of using AppState().modules
- Renamed PtActions.epoch_num to PtActions.epoch
- Removed some leftover code with respect to TrainableNeuralModuleWrapper
- Created TrainingState which replaces the registered_modules inside of train()
- Moved callback functions from Action functions to inside of PtActions.train()
- Deleted `__get_pytorch_module` from Actions.

Minor:

- Added NeuralGraphs to Jasper_an4 example
- Added a NmTensorNameRegistry and the ability for users to name tensors.

Tasks:
- [x] Create a NmTensor Registry
- [x] Add tensor.register inside of ~Neural Module calls~ NmTensor.__init__()
- [x] Allow users to rename their tensors
- [x] Creating state during steps
    - [x] Resetting state after each step
    - [x] Pass state to callbacks
    - [x] Allow users to access tensors from state
- [x] Map state["loss"] to final_loss
- [x] Allow users to compute value of uncomputed tensors - there is a somewhat hacky working version atm
- [x] Clean up the code used to compute values of uncomputed tensors
- [x] Refactor SimpleLossLogger
- [x] Create function decorators
- [x] Add new TensorboardLogger, WandBLogger, and CheckpointCallback callbacks
- [x] Bug fixes
- [x] Deprecation warnings
- [x] Changelog
- [x] add tests
- Add documentation

### CODE REVIEW

File: examples/asr/jasper_an4_debug.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    #     get_tb_values=lambda x: [["loss", x[0]]],
+    #     tb_writer=nf.tb_writer,
+    # )
+
+    # loss.rename("test")
+    # train_callback = nemo.core.SimpleLossLogger(tensors_to_log=["test"])
+
+    # train_callback = nemo.core.SimpleLossLogger()
+    predictions.rename("test")
+    train_callback = nemo.core.SimpleLossLogger(tensors_to_log=["loss", "test"])
```
Comment (tkornuta-nvidia) [inline examples/asr/jasper_an4_debug.py:94]: just wondering - what is the advantage of passing "names" instead of NmTensors?

File: nemo/core/neural_factory.py [hunk 2]
```diff
@@ -36,10 +37,117 @@
 from ..utils import ExpManager
 from .callbacks import ActionCallback, EvaluatorCallback
 from .neural_types import *
+from nemo.utils.app_state import AppState
 from nemo.utils.decorators import deprecated
 
 logging = nemo.logging
 
+# def topological_sort_from_leaves(leaf_nmtensors, cached_training_state: TrainingState = None):
+def topological_sort_from_leaves(leaf_nmtensors, cached_training_state=None):
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:46]: ? this looks like a copy of inverse graph traversing from actions, am I right?

File: nemo/core/neural_types/nmtensor_registry.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+
+class NmTensorNameRegistry:
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_types/nmtensor_registry.py:18]: 1) I really do not understand why do we need this entity - each neural graph already records ALL the tensors in it, an through AppState you have access to all graphs (AppState().graphs -> dict of graphs)
2) There is ObjectRegistry class which a) creates weak references to objects and b) generates unique names
3) I really do not understand why do you need ALL Tensors, where ALL == tensors from all graphs. This has serious implications. I honestly think that a given callback should operate within a scope of a single graph only.

File: nemo/utils/neural_graph/object_registry.py [hunk 4]
```diff
@@ -125,7 +125,7 @@ def __eq__(self, other):
         """
             Checks if two registers have the same content.
 
-            Args:
```
Comment (tkornuta-nvidia) [inline nemo/utils/neural_graph/object_registry.py:128]: Thanks for the cleanups, interesting, I though that black is taking care of that kind of flaws...

File: nemo/core/neural_types/neural_type.py [hunk 5]
```diff
@@ -223,6 +223,7 @@ def __init__(self, producer, producer_args, output_port_name, ntype=None):
         self._step_number = AppState().active_graph.step_number
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_types/neural_type.py:223]: I didn't see the need to store which graph is the "owner" of a given tensor.

But how about adding this information to NmTensor? Seems very easy:
self._graph = AppState().active_graph.name
(+ property def graph(self))

So then accessing all tensor in a given "graph_a":
AppState().graphs["graph_a"].tensors_list

Moreover, this will also give you access to "graph execution engine", which, at its core, is graph.steps -> note that e.g. DL should be step 0 (in a "valid executable graph"), so you won't even need to copy the "inverse graph traverse" ;)

So injection of this single line seems to be quite powerful ;)

File: nemo/core/neural_factory.py [hunk 6]
```diff
@@ -36,10 +37,117 @@
...
+        top_sorted_modules.append((m[0], dict(m[1]), m[2]))
+        # Ensure that there is only one dataset in callchain
+        if i > 0 and isinstance(m[0], DataLayerNM):
+            raise ValueError("There were more than one DataLayer NeuralModule inside your DAG.")
+
+        # TODO
+        if cached_training_state and isinstance(m[0], DataLayerNM):
+            raise ValueError("Could not compute tensor from current cached training state.")
+
+    return top_sorted_modules
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:149]: I really want to emphasize that this code is obsolete. Please note that neural graphs have this tiny property called steps.

For example, for graph:
```
with NeuralGraph(operation_mode=OperationMode.training) as g0:
    x, t = dl()
    p = m2(x=x)
    lss = loss(predictions=p, target=t)
```

self.steps contains:
```
    * Steps (3):
        0. realfunctiondatalayer0
        1. taylornet0
        2. mseloss0
```

File: nemo/core/callbacks.py [hunk 7]
```diff
@@ -37,6 +37,60 @@
...
+    @property
+    def action(self):
+        """TODO remove"""
+        return self._action
+
+    @action.setter
+    def action(self, action_obj):
+        self._action = action_obj
+
+    def on_action_start(self, state):
```
Comment (okuchaiev) [inline nemo/core/callbacks.py:53]: I thought you proposed to have this set of "events"
on_train_start
on_epoch_start
on_optimizer_step_start
on_batch_start
on_batch_end
on_optimizer_step_stop
on_epoch_end
on_train_end

File: nemo/core/callbacks.py [hunk 8]
```diff
@@ -37,6 +37,60 @@
...
+                tensor = state["tensors"].get_tensor(tensor_key)
+                if tensor is None:
+                    tensor = state["tensors"].get_and_compute_tensor(tensor_key, self.action)
+                logging.info("%s: %s", tensor_key, tensor)
+                # except KeyError:
+                #     raise KeyError(f"{self} was passed {tensor_key} but the tensor was not found in the state_dict. "
+                #                    f"Current state tensors include {state['tensors'].tensor_list()}")
+
+
 class ActionCallback(ABC):
```
Comment (okuchaiev) [inline nemo/core/callbacks.py:94]: Is the plan going forward to only have one "base" callback - NeMoCallback?

File: examples/asr/jasper_an4_debug.py (hunk 1 above)
Comment (blisc) [reply examples/asr/jasper_an4_debug.py:94]: I have no good answer, except that I think strings are easier for users to work with inside of their own callback functions.
On the other hand, NmTensors would make creating this PR easier.

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (blisc) [reply nemo/core/neural_factory.py:46]: It has been removed from actions

File: nemo/core/neural_types/nmtensor_registry.py (hunk 3 above)
Comment (blisc) [reply nemo/core/neural_types/nmtensor_registry.py:18]: 1. I need a mapping from a user-defined user to a tensor.unique_name
2. As I have mentioned in your NeuralGraph PR, tensors already have a property called `unique_name` that guarantee uniqueness so there is no reason for ObjectRegistry
3. Well as we have not moved forward with NeuralGraphs inside of actions.py, there is no "currently running graph" that I can fetch tensors from.

File: nemo/core/neural_factory.py (hunk 6 above)
Comment (blisc) [reply nemo/core/neural_factory.py:149]: Again as actions is not based on NeuralGraphs yet, I see no reason why this would be helpful for me.

Comment (blisc) [reply nemo/core/neural_factory.py:149]: Since, our call DAG is not generated from NeuralGraphs inside of actions, I cannot use the step property.

File: nemo/core/callbacks.py (hunk 8 above)
Comment (blisc) [reply nemo/core/callbacks.py:94]: I think the current callback system will be backwards compatible with the old system. So we can have both NeMoCallback and a deprecated ActionCallback.

File: examples/asr/jasper_an4.py [hunk 9]
```diff
@@ -17,64 +17,68 @@
...
 
 def create_dags(model_config_file, vocab, args, nf):
 
-    # Create a data_layer for training.
-    data_layer = nemo_asr.AudioToTextDataLayer.import_from_config(
-        model_config_file,
-        "AudioToTextDataLayer_train",
-        overwrite_params={"manifest_filepath": args.train_dataset, "batch_size": args.batch_size},
-    )
+    with NeuralGraph() as g0:
```
Comment (tkornuta-nvidia) [inline examples/asr/jasper_an4.py:27]: Tip: you do not need to "instantiate" modules within the graph scope, generally it has no effect

File: examples/asr/jasper_an4.py [hunk 10]
```diff
@@ -17,64 +17,68 @@
...
+        audio, audio_len, transcript, transcript_len = data_layer()
+        processed, processed_len = data_preprocessor(input_signal=audio, length=audio_len)
+        encoded, encoded_len = jasper_encoder(audio_signal=processed, length=processed_len)
+        log_probs = jasper_decoder(encoder_output=encoded)
+        predictions = greedy_decoder(log_probs=log_probs)
+        loss = ctc_loss(
+            log_probs=log_probs, targets=transcript, input_length=encoded_len, target_length=transcript_len,
+        )
+
+        # Create an evaluation graph.
```
Comment (tkornuta-nvidia) [inline examples/asr/jasper_an4.py:73]: Shouldn't that be a second graph?

File: nemo/backends/pytorch/actions.py [hunk 11]
```diff
@@ -87,137 +89,72 @@ def __init__(
...
+        top_sorted_modules = topological_sort_from_leaves(hook)
 
         if not isinstance(top_sorted_modules[0][0], DataLayerNM):
             raise ValueError("The first module in your DAG was not a DataLayer NeuralModule.")
 
         tdataset = top_sorted_modules[0][0].dataset
 
-        # populate self.module_reference_table
         for m in top_sorted_modules:
             if m[0].factory is None and self._local_rank is not None:
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:146]: Minor comment: I think that the first part of this condition simply cannot be True now, ever - factory is always set now in NM __init__()

File: nemo/backends/pytorch/actions.py [hunk 12]
```diff
@@ -372,10 +309,10 @@ def __initialize_amp(
         if optim_level == Optimization.mxprO0:
             return optimizer
 
-        if len(self.modules) < 1:
+        if len(AppState().modules) < 1:
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:312]: 👍

File: nemo/backends/pytorch/actions.py [hunk 13]
```diff
@@ -1093,6 +1010,179 @@ def train(
...
+                            "optimizers": action.optimizers,
+                        }
+                    )
+
+                def restore_state_from(self, path):
+                    if os.path.isfile(path):
+                        # map_location could be cuda:<device_id> but cpu seems to be more
+                        # general since we are also saving step and epoch
+                        # load_state_dict should move the variables to the relevant device
+                        checkpoint = torch.load(path, map_location="cpu")
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:1159]: Can you consider making it "backend-agnostic" using: nemo.backends.load
like in here:
https://github.com/NVIDIA/NeMo/blob/d91f349e46facb3f0943e2a4341557102a5bfd52/nemo/core/neural_graph.py#L1028

You know, one "agnostic" step at a time ;)

File: nemo/backends/pytorch/actions.py [hunk 14]
```diff
@@ -1093,6 +1010,179 @@ def train(
...
+                    else:
+                        raise FileNotFoundError("Could not find checkpoint file: {0}".format(path))
+
+                def save_state_to(self, path):
+                    state = {
+                        "step": self["step"],
+                        "epoch": self["epoch"],
+                        "optimizer_state": [opt.state_dict() for opt in self["optimizers"]],
+                    }
+                    torch.save(state, path)
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:1181]: The same here: nemo.backends.save

File: nemo/backends/pytorch/actions.py [hunk 15]
```diff
@@ -1187,7 +1277,7 @@ def train(
         # callbacks setup
         if callbacks is not None:
             for callback in callbacks:
-                if not isinstance(callback, ActionCallback):
+                if not isinstance(callback, ActionCallback) and not isinstance(callback, NeMoCallback):
                     raise ValueError("A callback was received that was not a child of ActionCallback")
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:1281]: "... nor NeMoCallback" ?

File: nemo/backends/pytorch/actions.py [hunk 16]
```diff
@@ -1252,67 +1340,70 @@ def train(
...
-                                rank = torch.distributed.get_rank()
-                                group = rank // synced_batchnorm_groupsize
-                                group_rank_ids = range(
-                                    group * synced_batchnorm_groupsize, (group + 1) * synced_batchnorm_groupsize
+            # for train_iter in training_loop:
+            #     call_chain = train_iter[2]
+            #     for i in range(1, len(call_chain) - 1):
+            #         key = call_chain[i][0].unique_instance_id
+            #         pmodule = self.module_reference_table[key][1]
+            #         num_trainable_weights = self.module_reference_table[key][1].num_weights
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/actions.py:1348]: There is plenty of those commented blocks... Why are we leaving them? What is our intention here?

File: nemo/core/callbacks.py [hunk 17]
```diff
@@ -15,6 +15,22 @@
...
+#     "on_train_start",
+#     "on_train_end",
+#     "on_epoch_start",
+#     "on_epoch_end",
+#     "on_batch_start",
+#     "on_batch_end",
+#     "on_step_start",
+#     "on_step_end",
+# ]
+
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:33]: commented... why we are adding this at all?

File: nemo/core/callbacks.py [hunk 18]
```diff
@@ -35,6 +54,463 @@
...
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self.tb_writer.add_scalar('misc/epoch', state["epoch"], state["step"])
+                self.tb_writer.add_scalar('misc/epoch_time', epoch_time, state["step"])
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:273]: Why do we have this inside of a TensorboardLogger?

This looks like a good, tiny TimeLogger callback to me... How about _decoupling_ those two functionalities?

File: nemo/core/callbacks.py [hunk 19]
```diff
@@ -35,6 +54,463 @@
...
+
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self._wandb_log({"epoch": state["epoch"], "epoch_time": epoch_time})
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:353]: In here it makes sense, as we are sending this information to W&B...

Comment (tkornuta-nvidia) [reply nemo/core/callbacks.py:353]: Or maybe not... seems that I have a problem with understanding what kind of "granularity" of callbacks we expect...

I honestly believe the smaller the better, as then a user can come and cherry-pick only the callbacks that he needs...

File: nemo/core/callbacks.py [hunk 20]
```diff
@@ -35,6 +54,463 @@
...
+    def _wandb_log(self, tensors_logged):
+        if _WANDB_AVAILABLE:
+            wandb.log(tensors_logged, step=state["step"])
+
+
+class CheckpointCallback(NeMoCallback):
+    def __init__(
+        self,
+        folder: str,
+        load_from_folder: str = None,
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:364]: Why inside of callback that is supposed to save we also have an option to load...?

Should _loading_ modules/checkpoint be part of callback system at all?

File: nemo/core/callbacks.py [hunk 21]
```diff
@@ -35,6 +54,463 @@
...
+        self._force_load = force_load
+
+    def __save_to(self, path, state):
+        if state["global_rank"] is not None and state["global_rank"] != 0:
+            return
+        if not os.path.isdir(path):
+            logging.info(f"Creating {path} folder")
+            os.makedirs(path, exist_ok=True)
+        unique_mod_names = set()
+        for module in AppState().modules:
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:409]: I truly believe this callback should work in the scope of _a single graph_.

Why are we saving all the modules?

File: nemo/core/callbacks.py [hunk 22]
```diff
@@ -35,6 +54,463 @@
...
+                logging.warning(e)
+                logging.warning(
+                    f"Checkpoint folder {path} was present but nothing was restored. Continuing training from random "
+                    "initialization."
+                )
+                return
+
+            try:
+                trainer_checkpoints = get_checkpoint_from_dir(["trainer"], path)
+                state.restore_state_from(trainer_checkpoints[0])
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:471]: ok, I thin we should explain the format of this checkpoint

It seems that state (from StateWrapper?) is in [0], what else is down there...?

File: nemo/core/neural_factory.py [hunk 23]
```diff
@@ -28,18 +28,210 @@
...
+
+    args:
+        leaf_nmtensors (List[NmTensors]): The tensors to be computed
+        cached_training_state (TrainingState): A dictionary of already computed tensors.
+            Defaults to None meaning an empty cache.
+
+    returns:
+        top_sorted_modules: the callchain DAG
+    """
+    from nemo.backends.pytorch.nm import DataLayerNM  # TODO: Replace this with a backend agnostic data layer
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:130]: use module.type == datalayer?

File: nemo/core/neural_factory.py [hunk 24]
```diff
@@ -28,18 +28,210 @@
...
+        if tensor_value is None and compute:
+            nmtensor = AppState().tensor_names._nmtensor_uniname_dict[unique_name]
+            callchain = topological_sort_from_leaves([nmtensor], cached_training_state=self)
+            callchain.insert(0, ())
+            self._action.nm_graph_forward_pass(callchain, self.tensor_dict)
+            tensor_value = self.tensor_dict[unique_name]
+        return tensor_value
+
+
+def topological_sort_from_leaves(leaf_nmtensors: List[NmTensor], cached_training_state: TrainingState = None):
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:118]: Ok, really I don't get it. Why this code, that is totally unrelated with Neural(Module)Factory, landed in here??

File: nemo/core/neural_factory.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_factory.py:232]: I think you will generate exactly the same DAG, and end up with the same order... But let's take it out of the scope of this PR

File: nemo/core/neural_factory.py [hunk 25]
```diff
@@ -154,29 +344,29 @@ def infer(self, tensors: List[NmTensor]):
...
-        """
-        Saves current state such as step, epoch and optimizer parameters
-        Args:
-          path:
+    # @abstractmethod
+    # def save_state_to(self, path: str):
+    #     """
+    #     Saves current state such as step, epoch and optimizer parameters
+    #     Args:
+    #       path:
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:352]: This seems to be an ingerence in NF API. Following our official guidelines - shoudn't we make it @deprecated first?

File: nemo/core/neural_factory.py [hunk 26]
```diff
@@ -194,50 +384,120 @@ def create_optimizer(self, optimizer, things_to_optimize, optimizer_params):
...
+    # def _update_callbacks(
+    #     self, callbacks=None, registered_tensors=None, final_loss=None,
+    # ):
+    #     # if self.local_rank is None or self.local_rank == 0:
+    #     if callbacks is not None and isinstance(callbacks, List) and len(callbacks) > 0:
+    #         for callback in callbacks:
+    #             if isinstance(callback, ActionCallback):
+    #                 callback._registered_tensors = registered_tensors
+    #             else:  # For now, we can use the old callback function. In the future we should improve this
+    #                 self.training_state.tensor_dict["loss"] = final_loss
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:500]: I honestly believe that leaving hundreds of commented code is not a good practice. This is not a production-level code.

File: nemo/core/neural_factory.py [hunk 27]
```diff
@@ -438,86 +698,6 @@ def __name_import(name):
...
-            constructor = NeuralModuleFactory.__name_import(collection_path)
-            if name == "BERT" and pretrained is True:
-                params["pretrained"] = True
-
-        # TK: "placement" is not passed as parameter anymore.
-        # if "placement" not in params:
-        #    params["placement"] = self._placement
-        instance = constructor(**params)
-        return instance
-
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py:520]: Thank you. :)

File: nemo/core/neural_types/__init__.py [hunk 28]
```diff
@@ -19,3 +19,4 @@
 from nemo.core.neural_types.comparison import *
 from nemo.core.neural_types.elements import *
 from nemo.core.neural_types.neural_type import *
+from nemo.core.neural_types.nmtensor_registry import NmTensorNameRegistry
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_types/__init__.py:22]: Analogical comment that I got from @okuchaiev regarding the other registries: do we want the user to be able to access it?

If not, please remove it from the namespace.

File: nemo/core/neural_types/neural_type.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_types/neural_type.py:223]: Ok, clearly you don't need it. I will add it when I will be working on other graph actions. Closing.

File: nemo/core/neural_types/nmtensor_registry.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_types/nmtensor_registry.py:18]: Thanks for the reply. Closing.

Comment (tkornuta-nvidia) [inline nemo/core/neural_types/nmtensor_registry.py:18]: So any particular reason why those shouldn't be weak references?

File: nemo/core/neural_types/nmtensor_registry.py [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+        args:
+            tensor (NmTensor): The tensor to be registered.
+        """
+
+        # Check if object is already in a set.
+        if tensor.unique_name in self._nmtensor_uniname_dict:
+            pass
+
+        # Finally, add object to the set.
+        self._nmtensor_uniname_dict[tensor.unique_name] = tensor
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_types/nmtensor_registry.py:49]: This looks like a "strong reference" to tensor object...

File: nemo/utils/app_state.py [hunk 30]
```diff
@@ -47,6 +47,17 @@ def __init__(self, device=None):
         self._module_registry = ObjectRegistry("module")
         # Create graph manager (registry with some additional functionality).
         self._neural_graph_manager = NeuralGraphManager()
+        # Create NmTensor registry
+        self._nmtensor_name_registry = nemo.core.neural_types.NmTensorNameRegistry()
+
+    @property
+    def tensor_names(self):
```
Comment (tkornuta-nvidia) [inline nemo/utils/app_state.py:54]: ... And in here you suggest that this is only mapping from one name to the other.

So what is really TensorRegistry storing?

File: tests/unit/core/test_nemo_callbacks.py [hunk 31]
```diff
@@ -0,0 +1,212 @@
+# ! /usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 NVIDIA. All Rights Reserved.
```
Comment (tkornuta-nvidia) [inline tests/unit/core/test_nemo_callbacks.py:4]: 2020? ;)

File: tests/unit/core/test_nemo_callbacks.py [hunk 32]
```diff
@@ -0,0 +1,212 @@
...
+
+@pytest.mark.usefixtures("neural_factory")
+class TestNeMoCallbacks:
+    @pytest.fixture()
+    def clean_up(self):
+        yield
+        self.nf.reset_trainer()
+
+    @pytest.fixture()
+    def create_tensorboard_file(self):
```
Comment (tkornuta-nvidia) [inline tests/unit/core/test_nemo_callbacks.py:40]: please use the tmpdif fixture from pytest

File: tests/unit/core/test_nemo_callbacks.py [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+        trainable_module = TaylorNet(dim=4)
+        loss = MSELoss()
+
+        # Create the graph by connnecting the modules.
+        x, y = data_source()
+        y_pred = trainable_module(x=x)
+        loss_tensor = loss(predictions=y_pred, target=y)
+
+        # Mock up both std and stderr streams.
+        with logging.patch_stdout_handler(StringIO()) as std_out:
```
Comment (tkornuta-nvidia) [inline tests/unit/core/test_nemo_callbacks.py:58]: ok, now I finally understand why you added those methods to logging... but should they really be part of logging, not testing?

File: tests/unit/core/test_nemo_callbacks.py [hunk 34]
```diff
@@ -0,0 +1,212 @@
...
+
+        # Create the graph by connnecting the modules.
+        x, y = data_source()
+        y_pred = trainable_module(x=x)
+        loss_tensor = loss(predictions=y_pred, target=y)
+
+        epoch_step_counter = [0]
+        epoch_batch_counter = [0]
+
+        @on_step_end
```
Comment (tkornuta-nvidia) [inline tests/unit/core/test_nemo_callbacks.py:177]: wow! Ok, those decorators are nice!

I haven't got that from the NeMoCallbacks API... what those are good for?

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_factory.py]: but.. why you put it in here??

File: nemo/core/callbacks.py (hunk 7 above)
Comment (tkornuta-nvidia) [reply nemo/core/callbacks.py]: I am against on_train_*, on_optimizer_* - there are specific for training action, and we already got two other types of **major actions** aside of training...

I would still suggest to use on_iteration_* instead of on_step_*, but I can live with it - as long we all agree on that name ;)

File: nemo/core/neural_factory.py (hunk 23 above)
Comment (okuchaiev) [reply nemo/core/neural_factory.py:130]: +1 this check "use module.type == datalayer" would make it backend-agnostic

File: nemo/core/callbacks.py (hunk 17 above)
Comment (blisc) [reply nemo/core/callbacks.py:33]: I wanted to add this but there are too many `import *` that touch this file and access names which I do not want exposed. After we remove support for the old callback system, I plan on adding this back in.

File: nemo/core/neural_factory.py (hunk 25 above)
Comment (okuchaiev) [reply nemo/core/neural_factory.py:352]: I'd say that exposing save_state_to to the user was a mistake in the first place

File: nemo/core/neural_factory.py (hunk 23 above)
Comment (blisc) [reply nemo/core/neural_factory.py:130]: Moved `topological_sort_from_leaves` back into backends\pytorch\actions.py and removed this line.

File: nemo/core/neural_factory.py (hunk 24 above)
Comment (blisc) [reply nemo/core/neural_factory.py:118]: Moved topological_sort_from_leaves back into backends\pytorch\actions.py

File: nemo/core/neural_factory.py (hunk 26 above)
Comment (blisc) [reply nemo/core/neural_factory.py:500]: This has been removed

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (okuchaiev) [reply nemo/core/neural_factory.py]: It is now backend agnostic (when datalayer check is fixed). So here is a better place for it than pytorch/actions

File: tests/unit/core/test_nemo_callbacks.py (hunk 31 above)
Comment (blisc) [reply tests/unit/core/test_nemo_callbacks.py:4]: Yep, good point, updated

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_factory.py]: ? but... please correct me - but this code has **nothing** to do with NeuralModuleFactory...?

File: nemo/core/callbacks.py (hunk 20 above)
Comment (okuchaiev) [reply nemo/core/callbacks.py:364]: chekpointing callback can resume training correctly (e.g. including optimizer state)

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_factory.py]: maybe I am missing something, I will discuss it with @blisc  directly

File: nemo/core/neural_factory.py (hunk 23 above)
Comment (okuchaiev) [reply nemo/core/neural_factory.py:130]: why? IMHO it is better if this is backend agnostic. And backends\pytorch\actions.py  is kind of overcrowded already.

File: nemo/core/neural_types/__init__.py (hunk 28 above)
Comment (okuchaiev) [reply nemo/core/neural_types/__init__.py:22]: I thought this should be visible for users who write their own callbacks. Did I get it wrong?

File: nemo/core/neural_factory.py [hunk 35]
```diff
@@ -84,6 +192,44 @@ class DeviceType(Enum):
...
+            callchain = topological_sort_from_leaves([nmtensor], cached_training_state=self)
+            # print(callchain)
+            callchain.insert(0, ())
+            action.nm_graph_forward_pass(callchain, self.tensor_dict)
+            # print(self.tensor_dict[unique_name])
+            tensor_value = self.tensor_dict[unique_name]
+        return tensor_value
+
+
 class Actions(ABC):
```
Comment (tkornuta-nvidia) [inline nemo/core/neural_factory.py]: Discussed with @blisc : move Actions + TrainingState + graph traversing-related code to a separate file

File: nemo/core/callbacks.py (hunk 17 above)
Comment (tkornuta-nvidia) [reply nemo/core/callbacks.py:33]: @blisc Split all the new/old callbacks into different files ;)

File: nemo/core/neural_types/nmtensor_registry.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_types/nmtensor_registry.py:18]: if you store names only - not important anymore

File: nemo/core/neural_factory.py (hunk 23 above)
Comment (blisc) [reply nemo/core/neural_factory.py:130]: Please see https://github.com/NVIDIA/NeMo/pull/615#discussion_r431445854

File: nemo/core/neural_types/__init__.py (hunk 28 above)
Comment (blisc) [reply nemo/core/neural_types/__init__.py:22]: Removed

File: nemo/core/neural_types/nmtensor_registry.py (hunk 3 above)
Comment (blisc) [reply nemo/core/neural_types/nmtensor_registry.py:18]: See https://github.com/NVIDIA/NeMo/pull/615#discussion_r431406913

File: nemo/core/neural_types/nmtensor_registry.py (hunk 29 above)
Comment (blisc) [reply nemo/core/neural_types/nmtensor_registry.py:49]: Removing this reference and switching _nmtensor_uniname_dict to be a set()

File: nemo/utils/app_state.py (hunk 30 above)
Comment (blisc) [reply nemo/utils/app_state.py:54]: Yep NmTensorRegistry is just a mapping of user's naming of nmtensors to their unique_names. It also keeps track of all unique_names so we can initialize the TrainingState object.

File: nemo/core/neural_factory.py (hunk 35 above)
Comment (blisc) [reply nemo/core/neural_factory.py]: Moved Actions, TrainingState and topological_sort_from_leaves to a new file called nemo/core/actions.py

File: tests/unit/core/test_nemo_callbacks.py (hunk 32 above)
Comment (blisc) [reply tests/unit/core/test_nemo_callbacks.py]: done

File: nemo/backends/pytorch/actions.py (hunk 16 above)
Comment (blisc) [reply nemo/backends/pytorch/actions.py]: Did a scrub over the files in this PR and removed commented code.

File: nemo/backends/pytorch/actions.py (hunk 15 above)
Comment (blisc) [reply nemo/backends/pytorch/actions.py:1239]: done

File: nemo/core/callbacks.py (hunk 17 above)
Comment (blisc) [reply nemo/core/callbacks.py:33]: done

File: nemo/core/actions.py [hunk 36]
```diff
@@ -0,0 +1,298 @@
+# ! /usr/bin/python
```
Comment (okuchaiev) [inline nemo/core/actions.py:1]: +1
